m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,745 @@
+
+import com.uber.m3.tally.BucketPair;
+import com.uber.m3.tally.BucketPairImpl;
+import com.uber.m3.tally.Buckets;
+import com.uber.m3.tally.CachedCounter;
+import com.uber.m3.tally.CachedGauge;
+import com.uber.m3.tally.CachedHistogram;
+import com.uber.m3.tally.CachedHistogramBucket;
+import com.uber.m3.tally.CachedStatsReporter;
+import com.uber.m3.tally.CachedTimer;
+import com.uber.m3.tally.Capabilities;
+import com.uber.m3.tally.CapableOf;
+import com.uber.m3.thrift.generated.CountValue;
+import com.uber.m3.thrift.generated.GaugeValue;
+import com.uber.m3.thrift.generated.M3;
+import com.uber.m3.thrift.generated.Metric;
+import com.uber.m3.thrift.generated.MetricBatch;
+import com.uber.m3.thrift.generated.MetricTag;
+import com.uber.m3.thrift.generated.MetricValue;
+import com.uber.m3.thrift.generated.TimerValue;
+import com.uber.m3.util.Duration;
+import com.uber.m3.util.ImmutableMap;
+import com.uber.m3.util.TCalcTransport;
+import com.uber.m3.util.TUdpClient;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TBinaryProtocol;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.protocol.TProtocolFactory;
+import org.apache.thrift.transport.TTransport;
+
+import java.net.InetAddress;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Phaser;
+
+/**
+ * An M3 reporter.
+ */
+public class M3Reporter implements CachedStatsReporter, AutoCloseable {
+    private static final int DEFAULT_METRIC_SIZE = 100;
+
+    private static final int DEFAULT_MAX_QUEUE_SIZE = 4096;
+    private static final int DEFAULT_MAX_PACKET_SIZE = 1440;
+
+    private static final int THREAD_POOL_SIZE = 10;
+
+    private static final String SERVICE_TAG = "service";
+    private static final String ENV_TAG = "env";
+    private static final String HOST_TAG = "host";
+
+    private static final String DEFAULT_HISTOGRAM_BUCKET_ID_NAME = "bucketid";
+    private static final String DEFAULT_HISTOGRAM_BUCKET_NAME = "bucket";
+    private static final int DEFAULT_HISTOGRAM_BUCKET_TAG_PRECISION = 6;
+
+    private static final int EMIT_METRIC_BATCH_OVERHEAD = 19;
+    private static final int MIN_METRIC_BUCKET_ID_TAG_LENGTH = 4;
+
+    private M3.Client client;
+
+    private final MetricBatch metricBatch;
+
+    private final Object calcLock = new Object();
+    private TCalcTransport calc;
+    private TProtocol calcProtocol;
+
+    private int freeBytes;
+
+    private ResourcePool resourcePool;
+
+    private String bucketIdTagName;
+    private String bucketTagName;
+    private String bucketValFmt;
+
+    // Holds metrics to be flushed
+    private BlockingQueue<SizedMetric> metricQueue;
```

Comment:
Why does the reporter need a queue & executors? Won't it be called periodically by the scope's scheduling service?

Reply:
Talked offline. It does need to be async as `metric.Report(...)` goes through the same path too.